rclpy/rclpy/__init__.py

```diff
@@ -61,11 +60,28 @@ def create_node(node_name):
 def spin_once(node, timeout_sec=None):
     wait_set = _rclpy.rclpy_get_zero_initialized_wait_set()
 
-    _rclpy.rclpy_wait_set_init(wait_set, len(node.subscriptions), 0, 0)
-
-    _rclpy.rclpy_wait_set_clear_subscriptions(wait_set)
-    for subscription in node.subscriptions:
-        _rclpy.rclpy_wait_set_add_subscription(wait_set, subscription.subscription_handle)
+    _rclpy.rclpy_wait_set_init(
+        wait_set,
+        len(node.subscriptions),
+        0,
+        0,
+        len(node.clients),
+        len(node.services))
+
+    for entity in ['subscription', 'client', 'service']:
+        _rclpy.rclpy_wait_set_clear_entities(entity, wait_set)
+        if entity == 'subscription':
+            for subscription in node.subscriptions:
+                _rclpy.rclpy_wait_set_add_entity(
+                    entity, wait_set, subscription.subscription_handle)
+        elif entity == 'client':
+            for client in node.clients:
+                _rclpy.rclpy_wait_set_add_entity(
+                    entity, wait_set, client.client_handle)
+        elif entity == 'service':
+            for service in node.services:
+                _rclpy.rclpy_wait_set_add_entity(
+                    entity, wait_set, service.service_handle)
```

Comment:
If there are five instances of this, then it may make sense to deduplicate as @dirk-thomas has suggested.